src/rubrix/labeling/text_classification/weak_labels.py

```diff
@@ -0,0 +1,395 @@
+
+    Examples:
+        Get the weak label matrix and a summary of the applied rules:
+
+        >>> def awesome_rule(record: TextClassificationRecord) -> str:
+        ...     return "Positive" if "awesome" in record.inputs["text"] else None
+        >>> weak_labels = WeakLabels(rules=[awesome_rule], dataset="my_dataset")
+        >>> weak_labels.matrix
+        >>> weak_labels.summary()
+
+        Use snorkel's LabelModel:
+
+        >>> from snorkel.labeling.model import LabelModel
+        >>> label_model = LabelModel()
+        >>> label_model.fit(L_train=weak_labels.train_matrix())
+        >>> label_model.score(L=weak_labels.test_matrix(), Y=weak_labels.annotation())
+        >>> label_model.predict(L=weak_labels.train_matrix())
+    """
+
+    def __init__(
+        self,
+        rules: List[Callable],
+        dataset: str,
+        ids: Optional[List[Union[int, str]]] = None,
+        query: Optional[str] = None,
+        label2int: Optional[Dict[str, int]] = None,
+    ):
+        self._rules = rules
+        self._dataset = dataset
+
+        # load records and check compatibility
+        self._records: List[TextClassificationRecord] = load(
+            dataset, query=query, ids=ids, as_pandas=False
+        )
+        if self._records[0].multi_label:
+            raise MultiLabelError(
+                "Multi-label text classification is not yet supported."
+            )
+
+        # apply rules -> create the weak label matrix, annotation array, final label2int mapping
+        self._matrix, self._annotation_array, self._label2int = self._apply_rules(
+            label2int
+        )
+
+    def _apply_rules(
+        self, label2int: Optional[Dict[str, int]]
+    ) -> Tuple[np.ndarray, np.ndarray, Dict[str, int]]:
+        """Apply the rules to the dataset.
+
+        Args:
+            label2int: An optional custom label2int mapping.
+
+        Returns:
+            The weak label matrix, the annotation array and their label2int mapping.
+
+        Raises:
+            MissingLabelError: When provided with a `label2int` dict, and a
+                weak label or annotation label is not present in its keys.
+        """
+        # call apply on the ElasticSearch rules
+        for rule in tqdm(self._rules, desc="Preparing rules"):
+            if isinstance(rule, Rule):
+                rule.apply(self._dataset)
+
+        # create weak label matrix, annotation array, final label2int
+        weak_label_matrix = np.empty(
+            (len(self._records), len(self._rules)), dtype=np.short
+        )
+        annotation_array = np.empty(len(self._records), dtype=np.short)
+        _label2int = {"None": -1} if label2int is None else label2int
+
+        for n, record in tqdm(
+            enumerate(self._records), total=len(self._records), desc="Applying rules"
+        ):
+            # First: fill annotation array
+            annotation = record.annotation or "None"
```

Comment:
Could be a problem define this default `"None"`  if some labelling function return `"None"`  as the matched label?

Reply:
Done, now the abstention label is `None`, not the string "None".